### PR TITLE
Fix SPIRVTile pass to handle linalg with scf.if

### PR DESCRIPTION
**Pull request by stevesuzuki-arm**
The existing logic to count scf.if ops for fast and slow path didn't handle the case where scf.if exists in linalg.generic, which should be excluded in the analysis for tiling

### CODE REVIEW

**Comment by stevesuzuki-arm** [comment]
Could you kick the workflow runs, please?

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp [hunk 1]
```diff
@@ -46,11 +46,19 @@ namespace iree_compiler {
...
-  // If there are `scf.if` ops, we have both a fast and slow paths for
-  // padding handling. Then we need to scan both regions to discover such
-  // computation ops so that we can tile and fuse both regions.
+  // If there are `scf.if` ops which have linalg ops, we have both a fast and
+  // slow paths for padding handling. Then we need to scan both regions to
+  // discover such computation ops so that we can tile and fuse both regions.
   SmallVector<scf::IfOp, 1> ifOps;
-  funcOp.walk([&ifOps](scf::IfOp ifOp) { ifOps.push_back(ifOp); });
+  funcOp.walk<WalkOrder::PreOrder>([&ifOps](Operation *op) -> WalkResult {
+    if (isa<linalg::LinalgOp, TilingInterface>(op)) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:54]: Could you just check in the walk if the parent is a `LinalgOp` ?

File: compiler/src/iree/compiler/Codegen/SPIRV/test/tile_linalg_ops.mlir [hunk 2]
```diff
@@ -52,3 +52,57 @@ func.func @innermost_reduction() {
...
+//  CHECK-DAG:  %[[C4:.+]] = arith.constant 4 : index
+//  CHECK-DAG:  %[[C256:.+]] = arith.constant 256 : index
+
+//      CHECK: scf.for
+//      CHECK:     scf.for %{{.+}} = %[[C0]] to %[[C256]] step %[[C4]]
+//      CHECK:       linalg.generic
+// CHECK-SAME:         ins(%{{.+}}slice : tensor<4xi32>)
+// CHECK-SAME:         outs(%{{.+}} : tensor<4xi32>)
+//      CHECK: scf.if
+//      CHECK: scf.if
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/test/tile_linalg_ops.mlir:108]: Additionally check `linalg.yield` at the end to guarantee that the above two scf.if is in the linalg.generic region?

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:54]: +1

**Review by antiagainst — CHANGES_REQUESTED** [review]
Thanks for fixing this! Interesting to see now we can have scf ops inside linalg op regions.

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp (hunk 1 above)
Comment (stevesuzuki-arm) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp]: Fixed, thanks.

File: compiler/src/iree/compiler/Codegen/SPIRV/test/tile_linalg_ops.mlir (hunk 2 above)
Comment (stevesuzuki-arm) [reply compiler/src/iree/compiler/Codegen/SPIRV/test/tile_linalg_ops.mlir:108]: Added, thanks.

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp]: I think the suggestion was to do something like

```c++
funcOp.walk<PreOrder>([&ifOps](scf::IfOp ifOp) {
  if (ifOp->getParentOfType<linalg::LinalgOp>()) // skip
  else // push and advance
});
```

Comment (stevesuzuki-arm) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVTile.cpp]: Thank you for the clarification. Now I think it has been fixed accordingly.

**Review by antiagainst — APPROVED** [review]

**Pull request merged**